### PR TITLE
Use per-page max compressed size estimate for compression

### LINKED ISSUE #10857: [BUG] Writing parquet can take too much memory
**Describe the bug**
I have some proprietary data that when I try to write it out to parquet the amount of GPU memory that is being allocated becomes really huge. In this case over 200 times the size of the input data. I have been debugging this trying to understand why it is happening.  I have it traced down to the output compression buffer when doing snappy compression.

This data set has hundreds of columns in it and many of them are arrays and structs and arrays of structs/etc. Most of the columns have relatively small page sizes and also a relatively small number of pages, but one of them for some reason I do not understand has a page size that is 4 times the size of the next biggest page and thousands of times large than most pages. The current algorithm calculates the output size based off of the largest input page * the total number of pages. Because of this skew it means that we are asking for at least 26 times more data than we need. The 26 times is an estimate by writing each of the columns out individually and looking at the max page size there.

**Steps/Code to reproduce bug**
Just read in the file and try to write it out again. I am happy to hand over the file out of band. It is relatively small.

**Expected behavior**
I would love it if we could make it so that this would not crash. I don't know if there is something off with how page sizes are calculated and we are just getting too much skew there, or if it would be better to bucket the size of the pages, or even just call the nvcomp API for each of the different page sizes to get a truly accurate allocation.

### PR DESCRIPTION
**Pull request by devavret**
Closes #10857 

The current behaviour of parquet writer is to get the estimate for maximum page compressed size by first finding the maximum page size and using nvcomp's `nvcompBatchedSnappyCompressGetMaxOutputChunkSize` API once for the largest page. The total output memory is allocated for max_compressed_page_size * num_pages.

This approach is pessimistic and over-allocates output buffer for batched compression.

This PR changes this to call `nvcompBatchedSnappyCompressGetMaxOutputChunkSize` for each page and sum up the result to get the output buffer size. This greatly reduces the peak memory consumption of parquet writer so the compression step is no longer the bottleneck.

### CODE REVIEW

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -431,6 +441,8 @@ __global__ void __launch_bounds__(128)
       ck_g.bfr_size           = page_offset;
       ck_g.page_headers_size  = page_headers_size;
       ck_g.max_page_data_size = max_page_data_size;
+      if (not comp_page_sizes.empty()) ck_g.compressed_size = comp_page_offset;
+      // printf("compressed_size %d\n", ck_g.compressed_size);
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:445]:
```suggestion
```

File: cpp/src/io/parquet/page_enc.cu [hunk 2]
```diff
@@ -1095,6 +1107,7 @@ __global__ void __launch_bounds__(128, 8)
     if (not comp_in.empty()) {
       comp_in[blockIdx.x]  = {base, actual_data_size};
       comp_out[blockIdx.x] = {s->page.compressed_data + s->page.max_hdr_size, compressed_bfr_size};
+      // printf("compressed_bfr_size: %d\n", compressed_bfr_size);
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:1110]:
```suggestion
```

File: cpp/src/io/parquet/writer_impl.cu [hunk 3]
```diff
@@ -82,6 +83,17 @@ parquet::Compression to_parquet_compression(compression_type compression)
...
+void print(rmm::device_uvector<T> const& d_vec, std::string label = "")
+{
+  std::vector<T> h_vec(d_vec.size());
+  cudaMemcpy(h_vec.data(), d_vec.data(), d_vec.size() * sizeof(T), cudaMemcpyDeviceToHost);
+  printf("%s (%lu)\t", label.c_str(), h_vec.size());
+  for (auto&& i : h_vec)
+    std::cout << i << " ";
+  printf("\n");
+}
+
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/writer_impl.cu:96]:
```suggestion
```

File: cpp/src/io/parquet/writer_impl.cu [hunk 4]
```diff
@@ -1038,6 +1106,7 @@ void snappy_compress(device_span<device_span<uint8_t const> const> comp_in,
                       comp_out.end(),
                       compressed_data_ptrs.begin(),
                       [] __device__(auto const& out) { return out.data(); });
+    // print(compressed_data_ptrs, "compressed_data_ptrs");
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/writer_impl.cu:1109]:
```suggestion
```

File: cpp/src/io/parquet/writer_impl.cu [hunk 5]
```diff
@@ -1048,7 +1117,7 @@ void snappy_compress(device_span<device_span<uint8_t const> const> comp_in,
                                                      compressed_bytes_written.data(),
                                                      nvcompBatchedSnappyDefaultOpts,
                                                      stream.value());
-
+    // print(compressed_bytes_written, "compressed_bytes_written");
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/writer_impl.cu:1120]:
```suggestion
```

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Looks good outside of some leftover debug code.

File: cpp/src/io/parquet/writer_impl.cu (hunk 3 above)
Comment (devavret) [reply cpp/src/io/parquet/writer_impl.cu:96]: I remember removing this. How did it get back in here?

File: cpp/src/io/parquet/page_enc.cu [hunk 6]
```diff
@@ -366,8 +369,9 @@ __global__ void __launch_bounds__(128)
             }
             page_g.max_hdr_size += stats_hdr_len;
           }
-          page_g.page_data        = ck_g.uncompressed_bfr + page_offset;
-          page_g.compressed_data  = ck_g.compressed_bfr + comp_page_offset;
+          page_g.page_data = ck_g.uncompressed_bfr + page_offset;
+          if (not comp_page_sizes.empty())
+            page_g.compressed_data = ck_g.compressed_bfr + comp_page_offset;
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:374]: Is it ok that `page_g.compressed_data` is only set if the page sizes isn't empty? This seems ok logically, but does something else expect it to be set to 0?

File: cpp/src/io/parquet/page_enc.cu [hunk 7]
```diff
@@ -431,6 +438,7 @@ __global__ void __launch_bounds__(128)
       ck_g.bfr_size           = page_offset;
       ck_g.page_headers_size  = page_headers_size;
       ck_g.max_page_data_size = max_page_data_size;
+      if (not comp_page_sizes.empty()) ck_g.compressed_size = comp_page_offset;
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:441]: Another place where I worry about uninitialized data. Do we need to set this to 0 if we don't have page sizes?

**Review by hyperbolic2346 — COMMENTED** [review]
Just some questions about this change. Looks good in general.

File: cpp/src/io/parquet/page_enc.cu (hunk 6 above)
Comment (devavret) [reply cpp/src/io/parquet/page_enc.cu:374]: This is a multi-pass kernel. In the pass where `comp_page_sizes` is empty, `pages` is also empty. So this has no effect.

File: cpp/src/io/parquet/page_enc.cu (hunk 7 above)
Comment (devavret) [reply cpp/src/io/parquet/page_enc.cu:441]: Same answer. https://github.com/rapidsai/cudf/pull/11066#discussion_r891672148

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by devavret** [comment]
rerun tests

File: cpp/src/io/parquet/writer_impl.cu [hunk 8]
```diff
@@ -1409,7 +1465,11 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
   }
 
   // Build chunk dictionaries and count pages
-  if (num_chunks != 0) { init_page_sizes(chunks, col_desc, num_columns); }
+  hostdevice_vector<size_type> comp_page_sizes;
+  if (num_chunks != 0) {
+    comp_page_sizes = init_page_sizes(
+      chunks, col_desc, num_columns, max_page_size_bytes, max_page_size_rows, stream);
+  }
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:1472]: [IILE](https://rigtorp.se/iife/#:~:text=In%20C%2B%2B%20a%20IIFE%20is,of%20statements%20into%20an%20expression.)?
Alternatively, could `init_page_sizes` handle empty input instead?

File: cpp/src/io/parquet/writer_impl.cu [hunk 9]
```diff
@@ -854,22 +855,77 @@ void writer::impl::gather_fragment_statistics(
...
+  // Get per-page max compressed size
+  hostdevice_vector<size_type> comp_page_sizes(0, num_pages, stream);
+  auto pages_comp_max_size = 0;
+  for (auto page_size : page_sizes) {
+    size_t page_comp_max_size = 0;
+    nvcompBatchedSnappyCompressGetMaxOutputChunkSize(
+      page_size, nvcompBatchedSnappyDefaultOpts, &page_comp_max_size);
+    pages_comp_max_size += page_comp_max_size;
+    comp_page_sizes.insert(page_comp_max_size);
+  }
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:912]: Could be made into a transform + accumulate

**Review by vuule — COMMENTED** [review]
What an interesting mess this turned into :)
Looks good (suggestions are not critical).

**Review by vuule — APPROVED** [review]

**Comment by devavret** [comment]
@gpucibot merge

**Pull request merged**